File: v1/runtime/backend/js/src/main/scala/scalascript/codegen/intrinsics/Payment.scala
package scalascript.codegen

import scalascript.backend.spi.*
import scalascript.ir.QualifiedName

/** Payment Request API intrinsics for the JS (browser/Node) backend.
 *
 *  Browser targets use the native `window.PaymentRequest` — the JS preamble
 *  emits `_prMethodData`, `_prDetails`, `_prOptions` helpers that map
 *  ScalaScript types to the W3C JSON format.  Node targets (server routes)
 *  emit calls to the JVM-equivalent runtime stubs so the same .ssc file
 *  compiles for both. */
val JsPaymentIntrinsics: Map[QualifiedName, IntrinsicImpl] = Map(
  // ── PaymentRequest lifecycle ────────────────────────────────────────────
  QualifiedName("PaymentRequest")                    -> RuntimeCall("_PaymentRequest"),
  QualifiedName("PaymentRequest.canMakePayment")     -> RuntimeCall("_pr_canMakePayment"),
  QualifiedName("PaymentRequest.show")               -> RuntimeCall("_pr_show"),
  QualifiedName("PaymentRequest.abort")              -> RuntimeCall("_pr_abort"),
  // ── PaymentResponse ──────────────────────────────────────────────────────
  QualifiedName("PaymentResponse.complete")          -> RuntimeCall("_pr_complete"),
  // ── Event hooks ──────────────────────────────────────────────────────────
  QualifiedName("PaymentRequest.onMerchantValidation")    -> RuntimeCall("_pr_onMerchantValidation"),
  QualifiedName("PaymentRequest.onShippingAddressChange") -> RuntimeCall("_pr_onShippingAddressChange"),
  QualifiedName("PaymentRequest.onShippingOptionChange")  -> RuntimeCall("_pr_onShippingOptionChange"),
  // ── Type constructors (mapped to plain JS objects) ───────────────────────
  QualifiedName("PaymentMethod.Card")                -> RuntimeCall("_PaymentMethodCard"),
  QualifiedName("PaymentMethod.ApplePay")            -> RuntimeCall("_PaymentMethodApplePay"),
  QualifiedName("PaymentMethod.GooglePay")           -> RuntimeCall("_PaymentMethodGooglePay"),
  QualifiedName("Amount")                            -> RuntimeCall("_Amount"),
  QualifiedName("PaymentItem")                       -> RuntimeCall("_PaymentItem"),
)

File: v1/runtime/backend/js/src/main/resources/scalascript/js-runtime/payment.mjs

// ── v1.38 Payment Request API preamble ───────────────────────────────────────

// ── Type constructors ─────────────────────────────────────────────────────────

function _Amount(currency, value) {
  return { currency, value };
}

function _PaymentItem(label, amount, pending) {
  return { label, amount, pending: !!pending };
}

function _PaymentMethodCard(networks, types) {
  return { _type: "card", networks: networks || [], types: types || [] };
}

function _PaymentMethodApplePay(merchantId, merchantName, countryCode) {
  return { _type: "apple-pay", merchantId, merchantName, countryCode: countryCode || "US" };
}

function _PaymentMethodGooglePay(merchantId, merchantName, environment) {
  return { _type: "google-pay", merchantId, merchantName, environment: environment || "TEST" };
}

// ── W3C JSON helpers ──────────────────────────────────────────────────────────

function _prMethodData(methods) {
  return methods.map(function(m) {
    if (m._type === "card") {
      const data = {};
      if (m.networks && m.networks.length) data.supportedNetworks = m.networks.map(function(n) { return n.toLowerCase(); });
      if (m.types && m.types.length) data.supportedTypes = m.types.map(function(t) { return t.toLowerCase(); });
      return { supportedMethods: "basic-card", data };
    } else if (m._type === "apple-pay") {
      return {
        supportedMethods: "https://apple.com/apple-pay",
        data: {
          version: 3,
          merchantIdentifier: m.merchantId,
          merchantCapabilities: ["supports3DS"],
          supportedNetworks: ["visa", "masterCard", "amex"],
          countryCode: m.countryCode
        }
      };
    } else if (m._type === "google-pay") {
      return {
        supportedMethods: "https://google.com/pay",
        data: {
          environment: m.environment,
          apiVersion: 2,
          apiVersionMinor: 0,
          merchantInfo: { merchantId: m.merchantId, merchantName: m.merchantName },
          allowedPaymentMethods: [{
            type: "CARD",
            parameters: { allowedAuthMethods: ["PAN_ONLY", "CRYPTOGRAM_3DS"], allowedCardNetworks: ["AMEX", "DISCOVER", "MASTERCARD", "VISA"] },
            tokenizationSpecification: { type: "PAYMENT_GATEWAY", parameters: { gateway: "example" } }
          }]
        }
      };
    }
    return { supportedMethods: m._type };
  });
}

function _prDetails(total, items, shippingOptions) {
  const details = { total: { label: total.label, amount: { currency: total.amount.currency, value: total.amount.value } } };
  if (items && items.length) {
    details.displayItems = items.map(function(i) {
      return { label: i.label, amount: { currency: i.amount.currency, value: i.amount.value }, pending: !!i.pending };
    });
  }
  if (shippingOptions && shippingOptions.length) {
    details.shippingOptions = shippingOptions.map(function(s) {
      return { id: s.id, label: s.label, amount: { currency: s.amount.currency, value: s.amount.value }, selected: !!s.selected };
    });
  }
  return details;
}

function _prOptions(opts) {
  if (!opts) return undefined;
  return {
    requestPayerName:  !!opts.requestPayerName,
    requestPayerEmail: !!opts.requestPayerEmail,
    requestPayerPhone: !!opts.requestPayerPhone,
    requestShipping:   !!opts.requestShipping,
    shippingType:      (opts.shippingType || "shipping").toLowerCase()
  };
}

// ── PaymentRequest wrapper ────────────────────────────────────────────────────

function _PaymentRequest(methods, total, items, opts, shippingOpts) {
  const methodData = _prMethodData(methods);
  const details    = _prDetails(total, items, shippingOpts);
  const options    = _prOptions(opts);
  const pr         = options ? new PaymentRequest(methodData, details, options)
                              : new PaymentRequest(methodData, details);
  pr._sscMethods = methods;
  return pr;
}

function _pr_canMakePayment(pr) {
  return pr.canMakePayment();
}

function _pr_show(pr) {
  return pr.show().then(function(resp) {
    resp._sscMethodName = resp.methodName;
    return resp;
  });
}

function _pr_abort(pr) {
  pr.abort();
}

function _pr_complete(resp, result) {
  const status = result === "Success" ? "success" : result === "Fail" ? "fail" : "unknown";
  return resp.complete(status);
}

function _pr_onMerchantValidation(pr, handler) {
  pr.addEventListener("merchantvalidation", function(e) {
    const p = handler(e);
    if (p && typeof p.then === "function") {
      p.then(function(session) { e.complete(session); });
    }
  });
}

function _pr_onShippingAddressChange(pr, handler) {
  pr.addEventListener("shippingaddresschange", function(e) {
    e.updateWith(handler(pr.shippingAddress));
  });
}

function _pr_onShippingOptionChange(pr, handler) {
  pr.addEventListener("shippingoptionchange", function(e) {
    e.updateWith(handler(pr.shippingOption));
  });
}
